### ISSUE TITLE
Apple Mail keeps re-downloading all of my emails

### ISSUE THREAD

**Issue by szlend** [ProtonMail/proton-bridge#29]
This has been a long standing issue for me and a number of people who've reported the issue over the years on Reddit and ProtonMail support. I ended up giving up and requesting a refund at the time. This year however I decided to give it another shot and see if the issue has been fixed. I'm sad to say that this is still a problem. I'm hoping now that the Bridge is open source, we can have an open conversation on why this is happening and how we can help finding the root cause.

I still haven't figured out the exact pattern of when this happens, but I started to write down my actions and I'll be updating this issue as it happens. It usually works fine for a few days, then after a while it starts re-downloading emails on a regular basis, sometimes several times per day. I can never find anything useful in the Bridge logs.
```
...
{"error":"mailbox  does not exist","level":"error","msg":"Could not get mailbox","name":"","pkg":"imap","time":"2020-04-28T11:56:23+02:00"}
{"error":"mailbox  does not exist","level":"error","msg":"Could not get mailbox","name":"","pkg":"imap","time":"2020-04-28T12:01:02+02:00"}
{"error":"mailbox  does not exist","level":"error","msg":"Could not get mailbox","name":"","pkg":"imap","time":"2020-04-28T12:02:04+02:00"}
{"error":"mailbox  does not exist","level":"error","msg":"Could not get mailbox","name":"","pkg":"imap","time":"2020-04-28T12:03:57+02:00"}
```

The last time this happened, I didn't do anything special. Just read a few emails, archive them, I un-flagged an email. A few minutes later I could hear my laptop fans spinning and I saw that the Bridge and Apple Mail were working full speed downloading my emails again.

What can we do to help you find the cause for this?

**Comment by horejsek** [comment]
Hi! We had problems with a sync (between Bridge and API). Sync was blocking any update (such as new message) and failed once user for example deleted or moved messages in web app. Fail was due to sync was not unable to cope with this and at the end once Bridge checked counts, it started again. We had to refactor Bridge a lot to fix this problem. 1.2.5 was huge improvement and ground for robust sync introduced in 1.2.6. Now sync between Bridge and API should be good.

Another problem is to follow precisely IMAP specification. Unfortunately this is not that easy. RFC is sometimes vague and each client implements it in it's own way. Besides mentioned sync we had issues that Bridge was not sending all updates resulting user did not see all messages. With our big refactor (Bridge 1.2.5) we tried our best to send all updates possible to not lose any message. It fixed a lot of problems. But introduced others. Last week we discovered why Outlook on Mac is rebuilding the whole mailbox once the message is moved from there. Apparently it's because we are sending both EXPUNGE with EXISTS to say message was removed from mailbox and stating new counts. Outlook on Mac did not expect EXISTS and re-builded mailbox just in case, causing constant re-sync (between Bridge and IMAP client). We have fix in QA stage. Once QA proves the fix, we will release new version and mirror the patch to this repository. It's possible Apple Mail has similar behaviour.

About constant `Could not get mailbox with name ""` in Apple Mail. We basically do not know what Apple Mail does that. RFC says any failed SELECT should result to unselected state but RFC is not saying we should accept `SELECT ""` as unselect and return OK. So we stick to RFC to return error. But we included IMAP extension UNSELECT and with that it seems Apple Mail is calling UNSELECT instead, hopefully fixing this issue. This is currently in QA as well.

Another problem we have currently with sync is localised error message. We had condition against  exact message because there was no other way which caused in special cases for non-English environment problem. This is also currently in QA.

> What can we do to help you find the cause for this?

Wait few days before we release mentioned fixes and see if it's still happening. First it's going to be beta. You could ask our support for beta access. If problem is still there, debug logs with as much information as possible are always welcome. 🙏

**Comment by exander77** [comment]
@horejsek 

> Sync was blocking any update (such as new message) and failed once user for example deleted or moved messages in web app. Fail was due to sync was not unable to cope with this and at the end once Bridge checked counts, it started again.

This seems to describe what I experienced in 1.2.6 Bridge.

I deleted the large number of messages in the Mail client and Web UI. And then bridge stopped receiving new emails and got 120% CPU: https://github.com/ProtonMail/proton-bridge/issues/25

**Comment by cuthix** [comment]
@szlend Do you have  `Download attachment` set to `All` (default is `Recent`)  in Apple Mail account settings?

**Comment by szlend** [comment]
@cuthix it's set to `Recent`, yes. I haven't touched the setting.

**Comment by telephon** [comment]
I have "solved" (a similar / this) issue in a slightly risky maneuver: In the protonmail web client, I've completely deleted all emails. Then I've restored them from my local backup via time machine. After that, the re-downloading stopped.

One issue I still have is that when I delete an email in my mail client, it gets re-downloaded from protonmail. This is also a bit annoying, but may be fixable by a setting perhaps?

**Comment by ciehanski** [comment]
I've been experiencing this issue ever since the release of Proton Bridge. My `Download attachment` is also set to `Recent`. Every week or other week Apple Mail goes through this process of re-downloading all my messages and monopolizes up my CPU and battery.

<img width="140" alt="Screen Shot 2020-05-15 at 20 31 40" src="https://user-images.githubusercontent.com/24802018/82107318-0b20e000-9716-11ea-8840-56de9516773b.png">

**Comment by telephon** [comment]
Recently, out of the blue, this happened here again, too.

**Comment by szlend** [comment]
Fyi, the Outlook fix in 1.2.7 didn't seem to have any effect on this on my end. It happened again just a few hours after updating.

**Comment by horejsek** [comment]
@szlend Thanks for the info. We found few more possible problems causing it. One quick improvement should come with 1.2.8 and others in future releases. We don't know when because it needs changes in go-imap library we are using. It will probably not be fully fixed before there is v2 of that library (https://github.com/emersion/go-imap/milestone/2, specifically because of EXPUNGE (https://github.com/emersion/go-imap/issues/323).

**Comment by szlend** [comment]
Given how much you rely on this library - this being one of the key differentiators to competing email service providers, I would expect ProtonMail to either sponsor the project or actively contribute to it in some meaningful way.

**Comment by telephon** [comment]
This is is still a problem in version 1.4.5. Note that this makes protonmail unusable in a mobile network connection. The data transfer happening is considerable, blocking other traffic.

<img width="290" alt="Screenshot 2020-10-25 at 10 00 42" src="https://user-images.githubusercontent.com/1810559/97102829-fe32f000-16a8-11eb-8541-a35ecfafac21.png">

So apparently, the problem is that having rules that sort incoming mail into separate folders causes protonmail to re-download them, and after this has happened, mail sorts them into the folders again, and ad infinitum (that is, until it hits some kind of limit that is set somewhere, I suppose).

**Comment by exander77** [comment]
@telephon The sorting rules you are talking about are in Apple Mail? And they operate on INBOX? Or do they operate on other folders as well? Like All Mail?

**Comment by exander77** [comment]
@szlend I am not a fan of Apple, but I have sympathy here. I think ProtonMail is extremely slow in debugging and fixing serious bugs. Go-imap v2 is in 9% progress, so this may take ages and somebody should do some workaround in the meantime.

**Comment by szlend** [comment]
@exander77 Yeah, the wait and see approach with go-imap v2 does not inspire much hope in me and there has been 0 visible progress on that milestone ever since this issue was opened. I was trying not to be too negative in my comments but honestly, the way this looks to me is like ProtonMail has no plan on actually fixing this and is merely hoping the open source community does it for them. The fact is that ProtonMail on Apple Mail remains hopelessly broken and has been in this state for years. Despite that, Apple Mails is still on the list of officially supported email clients. Personally I gave up on the bridge and am now using an Electron wrapper around the Web UI. Yeah, Thunderbird works better, but if I wanted to use a non-native looking application I might as well use the web client. The bridge was the sole reason I picked ProtonMail over some more affordable alternatives in the first place, so now I'm left wondering why even bother 😕

**Comment by telephon** [comment]
So how does Thunderbird solve this? It may be that the issue is on Apple's side.

>The fact is that ProtonMail on Apple Mail remains hopelessly broken and has been in this state for years. 

Well, this issue is the only one I have, the rest works fine.

**Comment by exander77** [comment]
@szlend For me, the Bridge is a critical feature as well. I choose to use ProtonMail because of this. So I understand that not be able to use it with the officially supported app is quite intolerable.

@telephon I have seen some bugs in mail clients. For example this Mailspring issue: https://github.com/ProtonMail/proton-bridge/issues/33 is completely Mailsping bug. Bur from the description this seems like a Bridge issue with the specific operation. You have to understand that IMAP is quite flexible Protocol and there are several ways how to do things. Apple Mail may do something differently than Thunderbird. But I doubt they are doing it wrong. I would have to debug the issue to see what is going on during the communication. Also with such a huge client like Apple Mail, they should look into it themselves.

**Comment by andrzejsza** [comment]
Could you please confirm if you still have the sync issues on 1.6.3? We've been doing a lot of work related to this in the meantime (from improving FETCH commands to catching body structures to reduce network traffic or improving communication with the API), so the issue should be much less severe now.
We're still looking again into parsing and better handling of poor/choppy connections.

**Comment by telephon** [comment]
Since the latest update to 1.6.3, I'm watching the situation, and so far I've not had the problem any more. Will post here.

**Comment by telephon** [comment]
Unfortunately, it persists:

<img width="284" alt="Screenshot 2021-02-19 at 12 05 26" src="https://user-images.githubusercontent.com/1810559/108497316-8fa71d00-72ab-11eb-8117-6a6ffe4504cf.png">

On macOS 10.15.7 (19H524)

**Comment by telephon** [comment]
And it gets recursively worse: 

<img width="283" alt="Screenshot 2021-02-20 at 14 46 57" src="https://user-images.githubusercontent.com/1810559/108597704-b7b97d80-738a-11eb-99e6-19d003f504c9.png">

**Comment by andrzejsza** [comment]
Could you contact us at bridge@protonmail.ch? We'd like to debug it more, but would need more info from you, if you don't mind. Also, keep in mind that the number shown client-side is misleading - for instance, all the messages are multiplied by the number of folders/labels they are in.

**Comment by telephon** [comment]
OK, I've done this, thanks @andrzejsza !

**Comment by kobiakov** [comment]
Hey @andrzejsza, am I right to assume that this issue is a duplicate of #130? Thanks!

**Comment by telephon** [comment]
Btw. As far as I am concerned, I had to get rid of the problem by deleting all my old mails from the server (a bit tedious, because the interface is limited to a bunch only).

**Comment by andrzejsza** [comment]
@kobiakov kind of. once the resync problem is sorted, we will be able to confirm if there are any other reasons for high CPU usage, so ok to keep them separately for now.

@telephon thanks so much for confirming. yes, this is definitely not something we can rely on as a solution but good to know that the problem is not affecting you anymore.

**Comment by telephon** [comment]
Currently I have 1840 mails in my "All Mail" folder. Suddenly the problem has started again, Proton Mail downloads over 5000 emails, and each round they are getting more.

**Comment by andrzejsza** [comment]
there were a lot of significant changes in 1.8.x that were aimed at mitigating the problem - namely the new message builder and the connection manager.
@telephon  if the problem persists (i.e. the resyc didn't finish and started to recur) please get in touch with support as we would like to have a look at the logs.

PSA - we know that some users are still hitting this problem. to all of you who do - please update to 1.8.x and keep us posted.
We are now working on improvements to local cache (so that we can serve client's requests faster).

**Comment by kobiakov** [comment]
I am running version 1.8.3 of the Bridge. 
Unfortunately, there little to no improvement with the re-downloading issue, it is happening for me right now.

**Comment by TXRed3** [comment]
Still a problem with the latest version. 
<img width="137" alt="image" src="https://user-images.githubusercontent.com/94854955/142901160-52ff63be-b3b8-4cc3-b019-77fa4fcd2ad2.png">

**Comment by angristan** [comment]
Still having the issue with v2.1

**Comment by andrzejsza** [comment]
@angristan how does the issue manifest for you with 2.1? is the initial sync completed and after that all the emails are re-downloaded for no obvious reason?

**Comment by angristan** [comment]
Yes, the initial sync went fine and almost every time I open Mail.app, it redownloads everything. No issues with Thunderbird.

**Comment by jat73** [comment]
Wanted to give this topic a bump. I'm on Bridge v2.1.1 and I've been a paying user, using Bridge with Apple Mail since the first Bridge beta. But the problem with constantly re-downloading emails (>78.000 in my case) and 100% CPU usage never went away. 
I have cleared the cache, removed all mails, re-installed the account, and tried everything I could find online, but nothing helped. 
So I am really really hoping this issue will be resolved at some point. As it is now, it's still pretty unusable and I don't want to use Thunderbird or Outlook as I have multiple accounts with other services and my own mail server and everything works fine with Apple Mail except for Proton. It's a shame it never got fixed, just seems it's not a big priority.

**Comment by tomtastic** [comment]
Same here, with just a 600MB ProtonMail mailbox, the initial Bridge setup and sync in Apple Mail has seen >12GB of data being downloaded by Bridge so far. It's been monopolising my internet connection for hours!

**Comment by telephon** [comment]
After updating macOS to Monterey, 12.4, the problem persists and I have the impression it has become a little more severe. The protonmail bridge luckily prevents the download of all these thousands of emails. But the updating process itself nevertheless costs a lot of energy, in bursts that happen every two hours and heat up the computer. Apple mail currently has the highest energy footprint of any app in my system for this reason.

**Comment by andrzejsza** [comment]
hi. just an update here - we've just opensourced our new go-mailserver https://github.com/ProtonMail/go-mailserver. this is still WIP but we'll start integrating into Bridge shortly.
we believe once this is done, it will be a permanent fix for two of the others biggest problems that we're facing:
https://github.com/ProtonMail/proton-bridge/issues/130
https://github.com/ProtonMail/proton-bridge/issues/220
any comments are welcome.

**Comment by telephon** [comment]
@andrzejsza this is good news! Is there a roadmap for the integration? (desperately waiting on an overheating computer …)

**Comment by andrzejsza** [comment]
yes, there is, and the integration is already in progress. very high level summary of what we're doing next:
- Implement the Proton connector
- Implement Synchronisation
- Implement EventLoop
- Provide file ownership to Gluon
- Start Gluon server on Bridge startup
- for release:
-- Remove dynamic feature flag
-- Remove cache on disk on bridge
-- Remove imap parsing/logic from bridge
-- Remove imap user management from bridge
in parallel, we're working on optimizing the new db.

as soon as we have the first beta, we'll share it here and in other relevant open issues.

**Comment by CyrilWendl** [comment]
Hi there

I just wanted to comment that I'm experiencing the same issues that have been described. Is there a timeline for this fix? Integration with Apple Mail is one of the features that has most value to me as I got used to the look and feel of Apple Mail over the years. Thank you!

**Comment by andrzejsza** [comment]
yes, there are two separate developments to mitigate this:
- nativity for M1 (in case this is what you're using, as per https://github.com/ProtonMail/proton-bridge/issues/155) - this will be release to beta next week
- switching to our own IMAP library, which will be included in Bridge V3 as per https://github.com/ProtonMail/proton-bridge/issues/220#issuecomment-1254315208 - aim for the beta is by the end of the year

**Comment by andrzejsza** [comment]
not many mew comments here, but I will strongly encourage all of you to try the new beta and let us know if you still have re-sync issues.

**Comment by jat73** [comment]
I updated to the latest Beta v3.0.8 on my MacBook Pro M1 today, reset everything, cleared the cache, and reconfigured Apple Mail. After re-downloading all of my 200k emails, it seems the problem has indeed been fixed. So far, I don't see any re-downloading of mails. Fingers crossed it will stay that way. Also, Bridge CPU usage is down to 0.2% from 100% and 140MB RAM, very nice.

**Comment by ylabbe** [comment]
Switching to beta 3.0.16 also fixed a related issue. When configuring my Apple Mail app with proton, the app would stay stuck downloading emails, completely breaking all functionality of the mail box. Thanks for the fix!

**Comment by andrzejsza** [comment]
indeed - v3 is fixing the problem. for any new sync-related bugs, feel free to open a new issue. closing.
